auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManager.java

```diff
@@ -100,34 +93,65 @@ public void saveCredentials(@NonNull Credentials credentials) {
         String accessToken = storage.retrieveString(KEY_ACCESS_TOKEN);
         final String refreshToken = storage.retrieveString(KEY_REFRESH_TOKEN);
         String idToken = storage.retrieveString(KEY_ID_TOKEN);
         String tokenType = storage.retrieveString(KEY_TOKEN_TYPE);
         Long expiresAt = storage.retrieveLong(KEY_EXPIRES_AT);
-        String scope = storage.retrieveString(KEY_SCOPE);
+        String storedScope = storage.retrieveString(KEY_SCOPE);
         Long cacheExpiresAt = storage.retrieveLong(KEY_CACHE_EXPIRES_AT);
         if (cacheExpiresAt == null) {
             cacheExpiresAt = expiresAt;
         }
 
-        if (isEmpty(accessToken) && isEmpty(idToken) || expiresAt == null) {
+        boolean hasEmptyCredentials = isEmpty(accessToken) && isEmpty(idToken) || expiresAt == null;
+        if (hasEmptyCredentials) {
             callback.onFailure(new CredentialsManagerException("No Credentials were previously set."));
             return;
         }
-        if (cacheExpiresAt > getCurrentTimeInMillis()) {
-            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), scope));
+
+        boolean willExpire = willExpire(cacheExpiresAt, minTtl);
+        boolean scopeChanged = hasScopeChanged(storedScope, scope);
+
+        if (!willExpire && !scopeChanged) {
+            callback.onSuccess(recreateCredentials(idToken, accessToken, tokenType, refreshToken, new Date(expiresAt), storedScope));
             return;
         }
         if (refreshToken == null) {
             callback.onFailure(new CredentialsManagerException("Credentials have expired and no Refresh Token was available to renew them."));
             return;
         }
 
-        authClient.renewAuth(refreshToken).start(new AuthenticationCallback<Credentials>() {
+        final ParameterizableRequest<Credentials, AuthenticationException> request = authClient.renewAuth(refreshToken);
+        if (scope != null) {
+            request.addParameter("scope", scope);
+        }
+        request.start(new AuthenticationCallback<Credentials>() {
             @Override
             public void onSuccess(@Nullable Credentials fresh) {
+                long nextCacheExpiresAt = calculateExpiresAt(fresh);
+                boolean willExpire = willExpire(nextCacheExpiresAt, minTtl);
+                if (willExpire) {
+                    long tokenLifetime = (nextCacheExpiresAt - getCurrentTimeInMillis() - minTtl * 1000) / -1000;
+                    CredentialsManagerException wrongTtlException = new CredentialsManagerException(String.format("The lifetime of the renewed Access Token or Id Token (%d) is less than the minTTL requested (%d). Increase the 'Token Expiration' setting of your Auth0 API or the 'ID Token Expiration' of your Auth0 Application in the dashboard, or request a lower minTTL.", tokenLifetime, minTtl));
```

Comment:
Idem, the use case that brought the TTL renewal feature to life was about making sure the AT would have enough lifetime before sending it to a backend server. Shouldn't only the AT be considered here? Specially considering the developer is not expected to send the ID Token anywhere.

Reply:
No, that would break compatibility with the feature requested in the linked PR (see the other comment).

Reply:
It's possible to check the TTL of just the AT without breaking compatibility, if you also check that the ID Token has not expired yet. So: 
- Check that the AT and the ID Token have not expired yet
- Check that the AT will not expire within the TTL
That will keep the logic consistent with Auth0.swift and will not break compatibility with that feature request.

Reply:
After all, the existing behavior (that needs to be preserved) is about checking if the credentials already expired, while the new behavior is about checking if they will expire. So the new, additive behavior can be limited to the AT, while keeping the existing behavior unchanged.

Reply:
Fixed. Do you mind taking a look, please?

Reply:
Please also update the error message, it still references the ID Token.